docs/spec.md

```diff
@@ -140,240 +43,355 @@ stablehlo.func @main(
+FuncInput   ::= '%' ValueId ':' ValueType
+FuncOutputs ::= ['->' FuncOutput, {',' FuncOutput}]
+FuncOutput  ::= ValueType
+FuncBody    ::= {Op}
+```
 
-A StableHLO program is executed by providing input values to the `main` function
-and computing output values. Output values of a function are computed by
-executing the graph of ops rooted in the corresponding `return` op.
+**StableHLO functions** (which are also called **named functions**) have
+an identifier, inputs/outputs and a body. In the future, we are planning to
+introduce additional metadata for functions to achieve better compatibility
+with HLO ([#425](https://github.com/openxla/stablehlo/issues/425)).
 
-The execution order is implementation-defined, as long as ops are executed
-before their uses. Possible execution orders of the example program above are
-`%0` → `%1` → `%2` → `%3` → `%4` → `return` or `%3` → `%0` → `%1` → `%2` → `%4`
-→ `return`.
+### Identifiers
 
-More formally, a **StableHLO process** is a combination of:
-(1) a StableHLO program, (2) operation statuses (not executed yet,
-already executed), and (3) intermediate values that the process is working on.
-The process starts with input values to the `main` function, progresses through
-the graph of ops updating operation statuses and intermediate values and
-finishes with output values. Further formalization is TBD.
+```ebnf
+FuncId  ::= '@' letter {letter | digit}
+ValueId ::= '%' digit {digit}
+          | '%' letter {letter | digit}
+letter  ::= 'a' | ... | 'z' | 'A' | ... | 'Z' | '_'
+digit   ::= '0' | ... | '9'
+```
 
-### Parallel execution
+**StableHLO identifiers** are similar to identifiers in many programming
+languages, with two peculiarities: 1) all identifiers have sigils which
+distinguish different kinds of identifiers, 2) value identifiers can be
+completely numeric to simplify generation of StableHLO programs.
 
-StableHLO programs can be executed in parallel, organized into a 2D grid of
-`num_replicas` by `num_partitions` which both have type `ui32`.
+### Types
 
-In the **StableHLO grid**, `num_replicas * num_partitions` of StableHLO
-processes are executing at the same time. Each process has a unique
-`process_id = (replica_id, partition_id)`, where
-`replica_id ∊ replica_ids = [0, ..., num_replicas-1]` and
-`partition_id ∊ partition_ids = [0, ..., num_partitions-1]` which both have
-type `ui32`.
+```ebnf
+Type         ::= ValueType | NonValueType
+ValueType    ::= TensorType | TokenType | TupleType
+NonValueType ::= ElementType | FunctionType | StringType
+```
 
-The size of the grid is known statically for every program, and the position
-within the grid is known statically for every process. Each process has access
-to its position within the grid via the `replica_id` and `partition_id` ops.
+**StableHLO types** are categorized into **value types** (which are also called
+**first-class types**) which represent StableHLO values and **non-value types**
```

Comment:
"StableHLO values"

Is the definition of a value: A type that can be an input or output of an Op ?